### PR TITLE
Add a clock to lab map

### PR DESCRIPTION
**Pull request by TonyLianLong**
I added a clock to the lab map.

Screenshot:

### CODE REVIEW

File: src/script.ts [hunk 1]
```diff
@@ -113,6 +113,17 @@ function enableNightMode(): void {
...
 }
 
+function updateClock(): void {
+    const m = new Date();
+    const clockTextElm = document.getElementById("clock-text");
+    if (clockTextElm == null) {
+        console.log("clock text element not found");
+    } else {
+        clockTextElm.textContent = (m.getHours() < 10 ? ("0" + m.getHours().toString()) : m.getHours())
+             + ":" + (m.getMinutes() < 10 ? ("0" + m.getMinutes().toString()) : m.getMinutes());
```
Comment (abizer) [inline src/script.ts]: what a shame JS doesn't have basic strftime support. in any case, prefer Freedom Units© because This Is America™

File: src/script.ts [hunk 2]
```diff
@@ -168,6 +179,10 @@ window.onload = function() {
     }
     setTimeout(location.reload, reloadTimer);
 
+    // Update the clock every second
+    setInterval(updateClock, 1000);
```
Comment (abizer) [inline src/script.ts:186]: no point in burning CPU updating every second if we're only displaying in minute precision

**Comment by raymondnumbergenerator** [comment]
thanks! I'm glad people are going through the effort of dealing with the mess I left behind

I would recommend positioning the clock a bit more evenly along the corner and have the size be consistent with the rest of the text on the page
having the ocf logo overlap the debian logo might also be a bit cluttered

I don't know what it looks like with also the regular hours displayed but I worry it might look a bit too cluttered with 3 different times listed on the page?

File: src/script.ts (hunk 2 above)
Comment (TonyLianLong) [reply src/script.ts:186]: Yeah, but if not every second what if we start at the 30th second of a minute. Probably a `setTimeout` so that `setInterval` can run at the first second of a minute?

File: src/script.ts (hunk 1 above)
Comment (TonyLianLong) [reply src/script.ts]: Isn't this already in customary units or do you mean the AM/PM stuff? @abizer

**Comment by TonyLianLong** [comment]
I can position the size of the clock right now, but as for size, I create an extra extra large text class because surely the purpose of adding this clock is to make people able to see the time from the OCF window (the wall clock is way too tiny).

For the OCF logo, I believe we can adjust the background color of it.

I can actually write another class to differentiate this from the OCF hours.

@raymondnumbergenerator

File: src/script.ts (hunk 2 above)
Comment (dkess) [reply src/script.ts:186]: Updating the clock every second barely burns CPU, considering we make an HTTP request every second anyways this seems like it would be a premature optimization. Second frequency is fine here.

File: www/night.scss [hunk 3]
```diff
@@ -5,6 +5,8 @@
 	.text-color-blue{fill:#5392D5;}
 	.text-size-medium{font-size:48px;}
 	.text-size-large{font-size:72px;}
+	.text-size-extra-large{font-size:144px;}
+	.text-size-extra-extra-large{font-size:216px;}
```
Comment (dkess) [inline www/night.scss:9]: I don't actually think it's necessary to specify this twice, since nightmode will only override, and not completely replace, daymode CSS. But since we are already duplicating all the other CSS perhaps this could go in a different PR

File: www/index.html [hunk 4]
```diff
@@ -1644,5 +1644,6 @@
     <text transform="matrix(1 0 0 1 1376.9951 1027.0986)" class="text-color text-font text-size-medium">Available Desktop :</text>
     <text transform="matrix(1 0 0 1 40.1084 97.6892)"><tspan id="labtext" x="0" y="0" class="text-color text-font text-size-large">Today’s Hours:</tspan>
     <tspan id="labhours" x="0" y="86.4" class="text-color text-font text-size-large"><tspan id="starthours"></tspan> - <tspan id="endhours"></tspan></tspan></text>
+    <text transform="matrix(1 0 0 1 100 1027.0986)" class="text-color text-font text-size-extra-extra-large" id="clock-text">00:00</text>
```
Comment (dkess) [inline www/index.html]: I'd initialize this as empty, since being frozen at `00:00` isn't correct.

File: src/script.ts (hunk 1 above)
Comment (dkess) [inline src/script.ts]: I think this could benefit from organization that makes it more clear what's going on. Something like (pseudocode):

```js
textContent = (
    m.getHours() < 10 ?  ...
    + ':'
    + m.getMinutes < 10 ? ...
)
```
Like where each part of the string is on its own line.

But also maybe we could use some TypeScript magic to get actual time formatting? That would be way better

**Review by dkess — APPROVED** [review]
Have some minor nits but otherwise lgtm, thanks for this!

**Comment by TonyLianLong** [comment]
Now it looks like this:
image

I found out that extra-extra-large is a bit too large after adding AM/PM so I tried a few sized and set them through `style` attribute.

File: src/script.ts [hunk 5]
```diff
@@ -113,6 +113,17 @@ function enableNightMode(): void {
...
 
+function updateClock(): void {
+    const time = new Date();
+    const clockTextElm = document.getElementById("clock-text");
+    if (clockTextElm == null) {
+        console.log("clock text element not found");
+    } else {
+        clockTextElm.textContent = time.toLocaleString("en-US", { hour: "numeric", minute: "numeric", hour12: true });
+    }
+    setTimeout(updateClock, (60 - time.getSeconds()) * 1000);
```
Comment (dkess) [inline src/script.ts]: The old way of updating this (with `setInterval`, outside of the function body) was better, it keeps the purpose of this function simpler.

I also still think updating this every second is fine but I won't die on that hill :P

**Comment by raymondnumbergenerator** [comment]
I think we should stick with the same font size as the rest of the text for the clock. Having different font sizes and colors feels very messy from a design perspective. Feel free to rename the classes for the text if you want. I think the way the lab hours is formatted should also stay the same so it doesn't get confusing with 3 different HH:MM times on the page.

Can you also try leaving the OCF logo in the bottom left like the current version and move the clock to the top right instead? The top right corner feels like a much more intuitive location for a digital clock on a screen. Feel free to move the debian logo lower or something if they overlap too much.

Here's another suggestion for a UI/UX improvement: having the ":" in the clock blink every second so it's more obvious that it's being updated.

I also agree with dkess regarding the second frequency.

**Comment by TonyLianLong** [comment]
Ok, I changed the frequency back then (which is also required for the blinking effect added right now). Also btw I didn't change the lab hours, I just modified it from developer tools and maybe it's not this format (maybe it's `9AM - 5PM` or something... I have no way to get the real lab hours locally). I moved OCF logo back and the debian logo a little lower. 

image

**Comment by raymondnumbergenerator** [comment]
that looks a lot better
imo the font color for the lab hours should stay the same too but lgtm otherwise now

**Comment by TonyLianLong** [comment]
Okay, now it's grey again. I'll merge it.
image

**Comment by bjb568** [comment]
Instead of `setInterval`, you can just `setTimeout` for the difference between now and the start of the next minute.

**Comment by TonyLianLong** [comment]
> Instead of `setInterval`, you can just `setTimeout` for the difference between now and the start of the next minute.

I did, and see the code review discussion by @dkess.

**Comment by bjb568** [comment]
I think it still makes sense to `setTimeout(updateClock, 1000 - new Date().getTime() % 1000)`.

**Comment by dkess** [comment]
May I ask why? As I explained earlier this makes the `updateClock` function's purpose simpler, and moves handling where and how it's called somewhere else, for a clean separation of concerns.

**Comment by bjb568** [comment]
1. It makes the time more accurate.
2. I don't think it's confusing because it's a common pattern in JS that I've seen before and is basically the async version of recursion. Maybe rename the function to `continuouslyUpdateClock` to make it clear it keeps going and calls itself more?

**Comment by dkess** [comment]
It doesn't make the time more accurate since it's still refreshing the clock every second (and either way, it's ok to be within one second of tolerance). And I'm still skeptical that this is good design, even if you may have seen it in other code.
